search-provider/yelp-search-provider.c

```diff
@@ -0,0 +1,568 @@
+                                         &error);
+
+    g_object_unref (session_bus);
+    if (reply)
+        g_variant_unref (reply);
+    if (error) {
+        g_dbus_method_invocation_return_gerror (invocation, error);
+        g_error_free (error);
+        return TRUE;
+    }
+    g_dbus_method_invocation_return_value (invocation, NULL);
+    return TRUE;
+
+}
+
+static gboolean 
+handle_launch_search (YelpShellSearchProvider2  *skeleton,
+                      GDBusMethodInvocation     *invocation,
+                      gchar                    **terms,
+                      guint32                    timestamp,
+                      gpointer                   user_data)
+{
+    gchar *joined_terms = g_strjoinv (" ", terms);
+    return yelp_launch (joined_terms, user_data, invocation, "show-search");
+}
+
+static gboolean
+handle_activate_result (YelpShellSearchProvider2  *skeleton,
+                        GDBusMethodInvocation     *invocation,
+                        gchar                     *result,
+                        gchar                    **terms,
+                        guint32                    timestamp,
+                        gpointer                   user_data)
+{
+    return yelp_launch (result, user_data, invocation, "show-page");
+}
+
+static void
+search_provider_app_dispose (GObject *obj)
+{
+    YelpSearchProviderApp *self = YELP_SEARCH_PROVIDER_APP (obj);
+
+    g_clear_object (&self->skeleton);
+    g_clear_pointer (&self->page_data_hash_map, g_hash_table_unref);
+
+    G_OBJECT_CLASS (yelp_search_provider_app_parent_class)->dispose (obj);
+}
+
+static void
+preload_data_cb (YelpDocument          *document,
+                 YelpDocumentSignal     signal,
+                 YelpSearchProviderApp *app,
+                 GError                *error)
+{
+    YelpSearchProviderApp *self = YELP_SEARCH_PROVIDER_APP (app);
+    gchar **page_ids = yelp_document_list_page_ids (document);
+    gchar **iter;
+
+    if (signal == YELP_DOCUMENT_SIGNAL_ERROR) {
+        if (error)
+            g_warning ("error: %s", error->message);
+        return;
+    }
+
+    if (signal == YELP_DOCUMENT_SIGNAL_CONTENTS)
+        return;
+
+    if (signal == YELP_DOCUMENT_SIGNAL_INFO) {
+        for (iter = page_ids; *iter; iter++) {
+            gchar *page_id = *iter;
+            PageData *data = g_new0 (PageData, 1);
+            gchar *icon_string = yelp_document_get_page_icon (document, page_id);
+            GIcon *icon = g_icon_new_for_string (icon_string, NULL);
+            g_free (icon_string);
+
+            data = page_data_new_steal (yelp_document_get_page_title (document, page_id), 
+                                        yelp_document_get_page_desc (document, page_id), 
+                                        icon);
+            g_hash_table_insert (app->page_data_hash_map, page_id, data);
+        }
+
+        g_application_release (G_APPLICATION (self));
+        g_free (page_ids);
```

Comment:
I think I asked for adding a comment here about stealing the contents of the `page_ids` array, so we free the array only here with `g_free` instead of the whole thing with `g_strfreev`.